src/main/kotlin/com/example/delivery/config/LogInvokeAnnotationBeanPostProccessor.kt

```diff
@@ -0,0 +1,88 @@
+import org.springframework.cglib.proxy.MethodInterceptor
+import org.springframework.core.Ordered
+import org.springframework.core.annotation.Order
+import org.springframework.stereotype.Component
+import java.lang.reflect.Method
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.memberProperties
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.isAccessible
+import kotlin.reflect.jvm.jvmErasure
+
+@Component
+@Order(Ordered.HIGHEST_PRECEDENCE)
+class LogInvokeAnnotationBeanPostProcessor : BeanPostProcessor {
+
+    private val beans = HashMap<String, AnnotationTarget>()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = bean.javaClass
+        when {
+            beanClass.isAnnotationPresent(InvokeLog::class.java) -> {
+                beans[beanName] = AnnotationTarget.CLASS
+            }
+
+            beanClass.declaredMethods.any { it.isAnnotationPresent(InvokeLog::class.java) } -> {
+                beans[beanName] = AnnotationTarget.FUNCTION
+            }
+        }
+        return bean
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any {
+        return if (beans.containsKey(beanName)) {
+            createProxy(bean, beanName)
+        } else {
+            bean
+        }
+    }
+
+    private fun createProxy(bean: Any, beanName: String): Any {
+        bean::class.memberProperties.forEach {
+            it.isAccessible = true
+        }
+
+        val constructorParams: List<KParameter> = bean::class.primaryConstructor!!.parameters
+        val argumentTypes: Array<Class<*>> = constructorParams.map { it.type.jvmErasure.java }.toTypedArray()
+        val arguments: Array<Any> = constructorParams.map { extractValue(it, bean) }.toTypedArray()
+
+        val enhancer = Enhancer()
+        enhancer.setSuperclass(bean.javaClass)
+        enhancer.setCallback(createInterceptor(beans.getValue(beanName)))
+
+        return enhancer.create(argumentTypes, arguments)
+    }
+
+    private fun extractValue(param: KParameter, bean: Any): Any {
+        return bean::class.memberProperties.find { it.name == param.name }!!.call(bean)!!
+    }
+
+    private fun createInterceptor(target: AnnotationTarget): MethodInterceptor {
+        return MethodInterceptor { obj, method, args, proxy ->
+            if (target == AnnotationTarget.CLASS || method.isAnnotationPresent(InvokeLog::class.java)) {
+                logMethod(method, args)
+            }
+            proxy.invokeSuper(obj, args)
+        }
+    }
+
+    private fun logMethod(method: Method, args: Array<out Any>) {
+        val className = method.declaringClass.name.substringAfterLast(".")
+        val message = method.parameters
+            .withIndex()
+            .joinToString(separator = ",") { (index, param) -> "${param.name} : ${args[index]}" }
+
+        log.info("Method called: {}.{}({})", className, method.name, message)
```

Comment:
We can make this code more efficient.

Try this code with different log level
```
LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
        .atDebug()
        .setMessage("Some expensive operation: {}")
        .addArgument {
            println("Expensive computation")
            "first arg"
        }
        .log()
```

So, the log message and parameters will be built only when required